### DISCUSSION TITLE
Can I make a correlation heatmap between taxa and categorical variable?

### DISCUSSION THREAD

**Discussion by timz0605** [david-barnett/microViz#163] [Q&A]
Hello David!

First of all, thank you for this amazing package!

I am trying to make some heatmaps to help visualize my data, and I am wondering if there are any ways to create correlation heatmaps between taxa and categorical variables? For example, my samples come from 4 sites (in metadata they are under the column `locality`), and each of the site could be classified as high, medium, or low impact. I was wondering if there is a way to create correlational heatmaps like that?

As a reference, below is the code I use to generate a composition heatmap:
```
taxa_heatmap <- ps_microviz %>%
  tax_agg("order") %>% 
  tax_transform(trans = "hellinger", rank = "order") %>%
  tax_filter(min_prevalence = 0.01, use_counts = TRUE) %>%
  comp_heatmap(
    colors = heat_palette(sym = TRUE), grid_col = NA,
    sample_side = "top", name = "Abd.",
    tax_anno = taxAnnotation(
      Prev. = anno_tax_prev(bar_width = 0.3, size = grid::unit(1, "cm"))),
    sample_anno = sampleAnnotation(
      Location = anno_sample("locality"),
      col = list(Location = cols), border = FALSE))
```

**Comment by david-barnett** [comment]
glad you're finding microViz useful 

To do this you need to recode your categorical data as numeric indicator variables. This is easy with `ps_mutate` and `if_else` or `case_when`

For nominal categorical variables like your `locality` variable.

```r
ps_microviz < - ps_microviz %>%
    ps_mutate(
          loc1 = if_else(locality == "placeA", true = 1L, false = 0L),
          loc2 = if_else(locality == "placeB", true = 1L, false = 0L),
          loc3 = if_else(locality == "placeC", true = 1L, false = 0L),
          loc4 = if_else(locality == "placeD", true = 1L, false = 0L)
    )
```

And for ordinal categorical variables like the high, med, low, `impact` you mentioned could either do the same thing or alternative create something like this

```r
ps_microviz < - ps_microviz %>%
    ps_mutate(impact_num = case_when(
        impact == "high" ~ 3,
        impact == "medium" ~ 2, 
        impact == "low" ~ 1
  ))
```

and then just use these for the correlation heatmap

**Comment by timz0605** [comment]
Hi David!

Thank you so much for your quick response!

Following up with that, I have a few questions about the annotation and aesthetics of the heatmap:

**1/ Is there a way to change/flip the hierarchical clustering of the variables? Or in other words, can I manually adjust the sequences of the variables in the heatmap? For example in this screenshot, I want to flip PI and TPC.**
 
![image](https://github.com/david-barnett/microViz/assets/114772844/e801cef3-7705-4f68-8555-af8704009c9d)

```
ps_microviz <- ps_microviz %>%
  ps_mutate(
    CI = if_else(locality == "Centre Island", true = 1L, false = 0L),
    CLP = if_else(locality == "Che Lei Pai", true = 1L, false = 0L),
    PI = if_else(locality == "Port Island", true = 1L, false = 0L),
    TPC = if_else(locality == "Tung Ping Chau", true = 1L, false = 0L)
  ) 
taxa_cor_heatmap <- ps_microviz%>% 
  tax_agg("order") %>% 
  tax_transform(trans = "hellinger", rank = "order") %>%
  tax_filter(min_prevalence = 0.01) %>%
  cor_heatmap(
    colors = heat_palette(palette = "Blue-Red 3"),
    vars = c("CI", "CLP", "PI", "TPC"),
    cor = "spearman"
    )
```

**2/ Is there a way to add a colored box legend to the correlation heatmap similar to what you did in the composition heatmap example?**

For example, in the screenshot below, I made a custom color scheme by specifying the color codes and storing them in `cols`. The code is the same as in my first post. I was looking at the reference and trying different options, but did not have much success. I've also looked into the examples you provided in the documentation page, but no such example was included...

![image](https://github.com/david-barnett/microViz/assets/114772844/90f15a3a-59cb-4ee4-b928-96884b038775)

**Comment by david-barnett** [comment]
**for 1:** try setting seriation_method_col ="Identity" to suppress the reordering of columns

**for 2:** I am not sure what you want a coloured box to indicate about each variable in a correlation heatmap, but it is possible to plot their univariate distributions, similar to the taxa: see the variable (var) annotation functions here - <https://david-barnett.github.io/microViz/reference/index.html#heatmap-plotting-functions>

**Reply by timz0605** [reply]
For 2, I am thinking of adding the colored location bar (that's currently showing on top of the abd. heatmap) also to the top of the correlation heatmap (could either replace the CI CLP TPC PI, or on top of the labels). Similar to the second legend in the abd. heatmap. 

For example:
![image](https://github.com/david-barnett/microViz/assets/114772844/9e0c1d09-28da-4849-bf2b-9224d6ef4e95)
![image](https://github.com/david-barnett/microViz/assets/114772844/cdedfc23-c0fe-4952-8710-f35a10c5228a)
I am using the same color scheme for these 4 locations in other figures I am making, so I was thinking this approach could help with connecting different figures. I was looking at `anno_cat`, but not sure if this is the right annotation function to use

**Reply by david-barnett** [reply]
ohhh, i did not notice they were the same variables, your colour-coding idea makes sense to me now.

you should be able to pass the argument `top_annotation` to `cor_heatmap`, and it will pass through to the call to `ComplexHeatmap::Heatmap`  

See this link for how to make a "simple annotation", which should allow what you're looking for I think:
https://jokergoo.github.io/ComplexHeatmap-reference/book/heatmap-annotations.html#heatmap-annotations
